cni-plugin/pkg/ipamplugin/ipam_plugin.go

```diff
@@ -252,6 +253,26 @@ func cmdAdd(args *skel.CmdArgs) error {
 			logrus.Info("Running in single-HNS-network mode, limiting number of IPAM blocks to 1.")
 			maxBlocks = 1
 		}
+		// Get namespace information for namespaceSelector support
+		namespace := epIDs.Namespace
+		namespaceLabels := map[string]string{}
+
+		// Only attempt to fetch namespace labels if we have Kubernetes configuration and a valid namespace
+		if (conf.Kubernetes.Kubeconfig != "" || conf.Policy.PolicyType == "k8s") && namespace != "" {
+			logger.Debugf("Getting namespace labels for: %s", namespace)
+
+			labels, err := getNamespaceLabels(calicoClient, namespace)
+			if err != nil {
+				logger.WithError(err).Warnf("Failed to get namespace labels for %s, using empty labels", namespace)
+				namespaceLabels = map[string]string{}
+			} else {
+				namespaceLabels = labels
+				logger.Debugf("Got namespace labels for %s: %v", namespace, namespaceLabels)
+			}
+		} else {
+			logger.Debugf("Using namespace %s without labels (no K8s config or empty namespace)", namespace)
+		}
+
 		assignArgs := ipam.AutoAssignArgs{
```

Comment:
We'll want to look for other places that call the IPAM assignment code and make sure they are given the Namespace information as well.

For example, the Service LoadBalancer controller likely wants this: https://github.com/projectcalico/calico/blob/291c8f5138429c8851bcbece1d5712fc1921567c/kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go#L674-L681

CC @MichalFupso

Reply:
https://github.com/projectcalico/calico/blob/291c8f5138429c8851bcbece1d5712fc1921567c/node/pkg/allocateip/allocateip.go#L477-L482

The tunnel allocation code probably doesn't care about Namespace labels - the tunnel IPs don't belong to a Namespaced resource. 

It does mean that any IP pool with a NamespaceSelector specified fundamentally cannot be used to assign tunnel IPs. We probably want validation logic that rejects IP pools that have `allowedUse` of `Tunnel` and that also have a NamespaceSelector specified.